### ISSUE TITLE
illegal reflective access in FileChannelDirectInputStream

### ISSUE THREAD

**Issue by jurgenvinju** [usethesource/vallang#136]
This warning will become an error if we go to Java 13:

```
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by io.usethesource.vallang.io.binary.util.FileChannelDirectInputStream (file:/home/runner/.m2/repository/org/rascalmpl/rascal/0.20.0-RC1/rascal-0.20.0-RC1.jar) to method java.nio.DirectByteBuffer.cleaner()
WARNING: Please consider reporting this to the maintainers of io.usethesource.vallang.io.binary.util.FileChannelDirectInputStream
```

**Comment by jurgenvinju** [comment]
```
// from: http://stackoverflow.com/a/19447758/11098
    private static void closeDirectBuffer(@Nullable ByteBuffer cb) {
        if (cb==null || !cb.isDirect()) return;

        // we could use this type cast and call functions without reflection code,
        // but static import from sun.* package is risky for non-SUN virtual machine.
        //try { ((sun.nio.ch.DirectBuffer)cb).cleaner().clean(); } catch (Exception ex) { }
        try {
            Method cleaner = cb.getClass().getMethod("cleaner");
            cleaner.setAccessible(true);
            Method clean = Class.forName("sun.misc.Cleaner").getMethod("clean");
            clean.setAccessible(true);
            Object param = cleaner.invoke(cb);
            if (param != null) {
                clean.invoke(param);   
            }
        } catch(Exception ex) { }
        cb = null;
    }
```

this is the offending code. But wouldn't a simple `System.gc()` do the trick as well? Why do we have to be sure to clean the memory mapped file here if the garbage collector is supposed to do this?

**Comment by jurgenvinju** [comment]
I understand `System.gc()` is not a guarantee, before we start discussing this. I only need to know why we are doing our own garbage collection here.

**Comment by DavyLandman** [comment]
duplicate of #119

**Comment by DavyLandman** [comment]
And no, it's not that simple. this is about closing file handles not about clearing memory. there are some discussion on this specific case. it's easy to solve in java 11, but we would have to adopt the java module system of java 11.

**Comment by jurgenvinju** [comment]
I don't understand what the SO answer is saying. All I see is reflective code to call an internal private method to unmap a memory mapped file. I want to understand why this is needed. Do you know? The SO answer explains why the reflection is needed to call the method, but not why the method call is needed in the first place.

According to the JVM's memory model we are not supposed to have to unmap memory mapped files. The GC is supposed to do this, or the internals of the code that does the memory mapping.

Also I would like to understand why the module system of Java 11 would help. Can you explain? If it's only to work around the reflection warning/error, then I'd like to work on the above issue first and try to avoid the need for unmapping/collecting file memory ourselves.

**Comment by DavyLandman** [comment]
> I don't understand what the SO answer is saying. All I see is reflective code to call an internal private method to unmap a memory mapped file. I want to understand why this is needed. Do you know? The SO answer explains why the reflection is needed to call the method, but not why the method call is needed in the first place.
>
> According to the JVM's memory model we are not supposed to have to unmap memory mapped files. The GC is supposed to do this, or the internals of the code that does the memory mapping.

The question explained the problem. It roughly goes like this: you open a memory map on a file, you `.close()` the memory-mapped backed ByteBuffer. However, if you later try to open the file using a regular file stream, you won't get the contents yet, or you'll get an exception that the file is still open and not closed yet.

This is caused by the `close` not actually releasing the handles back. The documentation of the [MappedByteBuffer](https://docs.oracle.com/javase/8/docs/api/java/nio/MappedByteBuffer.html) mentions this briefly: 

> A mapped byte buffer and the file mapping that it represents remain valid until the buffer itself is garbage-collected. 

So this mess with reflection (this kind of code can be found in most java programs that work with file channels) is needed to force the actual closing of the file handle.

> Also I would like to understand why the module system of Java 11 would help. Can you explain? If it's only to work around the reflection warning/error, then I'd like to work on the above issue first and try to avoid the need for unmapping/collecting file memory ourselves.

Yes, in java 11 (and required from java 13) to get access to the cleaner interface you need to have a module dependency on `jdk.unsupported`.


Also, shall we merge this with issue #119?

**Comment by DavyLandman** [comment]
Now when do we need file channel based access? It helps with reading large `IValue` files. We would have to profile how much again, but it was quite a bit back when I spent all the time adding this feature into both vallang and rascal.

**Comment by jurgenvinju** [comment]
Thanks for explaining. Sounds like a serious bug. Have they not fixed this in 13, 15 or 17 by now?

**Comment by jurgenvinju** [comment]
I suspect like you that file channels make a big difference

**Comment by jurgenvinju** [comment]
Closed because of duplicate
